File: velocity/src/main/java/gg/grounds/GroundsPluginServerDiscovery.java
package gg.grounds;

import com.google.inject.Inject;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.player.PlayerChooseInitialServerEvent;
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
import com.velocitypowered.api.event.proxy.ProxyShutdownEvent;
import com.velocitypowered.api.plugin.Plugin;
import com.velocitypowered.api.proxy.ProxyServer;
import gg.grounds.velocity.VelocityDiscoveryConfig;
import gg.grounds.velocity.VelocityDiscoveryService;
import org.slf4j.Logger;

@Plugin(id = "plugin-server-discovery", name = "Grounds Server Discovery Plugin")
public final class GroundsPluginServerDiscovery {
    private final ProxyServer proxyServer;
    private final Logger logger;
    private VelocityDiscoveryService discoveryService;

    @Inject
    public GroundsPluginServerDiscovery(
            ProxyServer proxyServer,
            Logger logger
    ) {
        this.proxyServer = proxyServer;
        this.logger = logger;
    }

    @Subscribe
    public void onProxyInitialize(ProxyInitializeEvent event) {
        proxyServer.getAllServers().forEach(server -> {
            logger.info("Removing pre-configured server: {}", server.getServerInfo().getName());
            proxyServer.unregisterServer(server.getServerInfo());
        });

        VelocityDiscoveryConfig config = VelocityDiscoveryConfig.load();

        discoveryService = new VelocityDiscoveryService(proxyServer, logger, config);
        discoveryService.start(this);
        logger.info("Velocity server discovery initialized");
    }

    @Subscribe
    public void onProxyShutdown(ProxyShutdownEvent event) {
        if (discoveryService != null) {
            discoveryService.stop();
        }
    }

    @Subscribe
    public void onPlayerChooseInitialServer(PlayerChooseInitialServerEvent event) {
        if (discoveryService == null || event.getInitialServer().isPresent()) {
            return;
        }
        discoveryService.pickInitialServer().ifPresent(event::setInitialServer);
    }
}

File: velocity/src/main/kotlin/gg/grounds/GroundsPluginServerDiscovery.kt
package gg.grounds

import com.google.inject.Inject
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.player.PlayerChooseInitialServerEvent
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent
import com.velocitypowered.api.event.proxy.ProxyShutdownEvent
import com.velocitypowered.api.plugin.Plugin
import com.velocitypowered.api.proxy.ProxyServer
import gg.grounds.discovery.VelocityDiscoveryConfig
import gg.grounds.discovery.VelocityDiscoveryService
import org.slf4j.Logger

@Plugin(id = "plugin-server-discovery", name = "Grounds Server Discovery Plugin")
class GroundsPluginServerDiscovery
@Inject
constructor(private val proxyServer: ProxyServer, private val logger: Logger) {
    private lateinit var discoveryService: VelocityDiscoveryService

    @Subscribe
    fun onProxyInitialize(event: ProxyInitializeEvent) {
        proxyServer.allServers.forEach { server ->
            logger.info("Removing pre-configured server: {}", server.serverInfo.name)
            proxyServer.unregisterServer(server.serverInfo)
        }

        val config = VelocityDiscoveryConfig.load()

        discoveryService =
            VelocityDiscoveryService(proxyServer, logger, config).also { service ->
                service.start(this)
            }
        logger.info("Velocity server discovery initialized")
    }

    @Subscribe
    fun onProxyShutdown(event: ProxyShutdownEvent) {
        if (this::discoveryService.isInitialized) {
            discoveryService.stop()
        }
    }

    @Subscribe
    fun onPlayerChooseInitialServer(event: PlayerChooseInitialServerEvent) {
        if (event.initialServer.isPresent) {
            return
        }
        discoveryService.pickInitialServer().ifPresent(event::setInitialServer)
    }
}
